app/javascript/components/tree-view/hierarchical-tree-view.jsx

```diff
@@ -1,53 +1,62 @@
 /* eslint camelcase: ["warn", {allow: ["bs_tree", "tree_name", "click_url", "check_url", "allow_reselect"]}] */
-import React, { useState } from 'react';
+import React, { useEffect } from 'react';
+import { connect } from 'react-redux';
 import PropTypes from 'prop-types';
-import Tree, { ActionTypes } from '@manageiq/react-ui-components/dist/wooden-tree';
+import Tree, {
+  Node,
+  defaultStore,
+  callBack,
+  ActionTypes,
+} from '@manageiq/react-ui-components/dist/wooden-tree';
 
 const convertData = data => Tree.convertHierarchicalTree(Tree.initHierarchicalTree(JSON.parse(data)));
 
 const HierarchicalTreeView = (props) => {
   const {
-    bs_tree, check_url, oncheck, checkboxes, allow_reselect,
+    tree_name,
+    bs_tree,
+    check_url,
+    oncheck,
+    checkboxes,
+    allow_reselect,
+    callBack,
   } = props;
-  const [tree, setTree] = useState(convertData(bs_tree));
 
-  const checked = (node, value) => {
-    const newNode = { ...node, state: { ...node.state, checked: value } };
-    if (oncheck) {
-      ManageIQ.tree.checkUrl = check_url;
-      window[oncheck](newNode);
-    }
-    return newNode;
-  };
+  const storeName = `tree_${tree_name}`;
+  const ConnectedNode = connect((store, ownProps) => ({ ...store[storeName][ownProps.nodeId] }))(Node);
+  const ReduxTree = connect(store => ({ data: { ...store[storeName] } }))(Tree);
 
-  const actionMapper = {
-    [ActionTypes.EXPANDED]: Tree.nodeExpanded,
-    [ActionTypes.CHECKED]: checked,
-    [ActionTypes.DISABLED]: Tree.nodeDisabled,
-    [ActionTypes.SELECTED]: Tree.nodeSelected,
-    [ActionTypes.CHILD_NODES]: Tree.nodeChildren,
-    [ActionTypes.LOADING]: Tree.nodeLoading,
-  };
+  useEffect(() => {
+    ManageIQ.redux.addReducer({ [storeName]: defaultStore });
+  }, []);
+
+  useEffect(() => {
+    callBack('', ActionTypes.ADD_NODES, convertData(bs_tree));
+  }, [bs_tree]);
 
   const onDataChange = (commands) => {
-    let NewTree = { ...tree };
     commands.forEach((command) => {
-      NewTree = (command.type === ActionTypes.ADD_NODES)
-        ? Tree.addNodes(NewTree, command.value)
-        : Tree.nodeUpdater(NewTree, actionMapper[command.type](Tree.nodeSelector(NewTree, command.nodeId), command.value));
+      callBack(command.nodeId, command.type, command.value);
+
+      // On change enables the buttons on the bottom of the form (or disables)
+      if (command.type === ActionTypes.CHECKED && oncheck) {
```

Comment:
This `onchange` might be different for each tree, so the comment above is invalid.

Reply:
I this is very clear. After component mounts it will add an reducer to store.